### PR TITLE
Adopt pytest as testing framework

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description

Adopting pytest as testing framework. This version of my code outputs an assertion error.

## Motivation and Context

After this change, We'll be using pytest instead of unittest to perform the unit tests.
## Types of Changes
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [ ] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [ ] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,3 +2,5 @@ signac>=1.0.0
 jinja2>=2.8
 cloudpickle
 deprecation>=2
+pytest>=3.5.0
+Click == 7.0
```
Comment (csadorf) [inline requirements.txt]: We need `Click` for the execution of the package.

**Comment by kidrahahjo** [comment]
> @kidrahahjo Thank you very much for the contribution, I have a few comments/questions:
> 
> * Did you use [unittest2pytest](https://pypi.org/project/unittest2pytest/) or a similar tool to make these changes?
> * When I look at the [CI output](https://circleci.com/gh/glotzerlab/signac-flow/6657?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), it appears that there is a massive amount of data dumped to STDOUT. Do you have any idea what the issue could be?

I used unittest2pytest for the changes. And about the massive log text appearance, I don't have much idea about it right now but I'm working on it and going through the code more thouroghly to know about that problem

**Comment by kidrahahjo** [comment]
<b>Facing this issue in test_project.py</b>
1. <i>Assertion Error (line 957)</i>
<b>Error Message</b>
AssertionError: assert 'Some of the keys provided as part of the directives were not used by the template script, including: bad_directive\n' in "Submitting cluster job 'FlowTestProject/bundle/5aa61ec57616198edbc19d0b39b7f35071f61550':\n - Operation: op1(017d53de...e34189b88)\n - Operation: op2(386b19932c82f3f9749dd6611e846293)\n - Operation: op2(5e4d14d82c320bafb2f1286fe486d1f8)\n"+ where "Submitting cluster job 'FlowTestProject/bundle/5aa61ec57616198edbc19d0b39b7f35071f61550':\n - Operation: op1(017d53de...e34189b88)\n - Operation: op2(386b19932c82f3f9749dd6611e846293)\n - Operation: op2(5e4d14d82c320bafb2f1286fe486d1f8)\n" = <built-in method getvalue of _io.StringIO object at 0x7fce641b61f8>() + where <built-in method getvalue of _io.StringIO object at 0x7fce641b61f8> = <_io.StringIO object at 0x7fce641b61f8>.getvalue

Screenshot from 2020-01-31 18-27-56


<b>Facing these issues in test_templates.py</b>
1. <i>Assertion Error (line 87)</i>
<b>Error Message</b>
AssertionError: assert '---------- S...ded78a54f68\n' == '---------- S...ded78a54f68\n' 
Skipping 3490 identical leading characters in diff, use -v to show

2. <i>Runtime Error len(jobs) returns 0 (line 39)</i>
<b>Error Message</b>
RuntimeError: No reference data for environment environments.umich.FluxEnvironment!

Screenshot from 2020-01-31 18-33-14

File: tests/test_environment.py [hunk 2]
```diff
@@ -8,29 +8,30 @@
 from flow.environment import TestEnvironment
 from flow.errors import ConfigKeyError
 from test_project import StringIO, redirect_stdout
+import pytest
 
 
 class ProjectTest(unittest.TestCase):
```
Comment (vishav1771) [inline tests/test_environment.py]:
``` suggestion
 class TestProject():

Comment (vishav1771) [reply tests/test_environment.py]: Instead of inheriting from `unittest.TestCase`  you can change the name of the class  starting from `Test`.

File: tests/test_environment.py [hunk 3]
```diff
@@ -8,29 +8,30 @@
...
         a = env.get_config_value('a', None)
-        self.assertIsNone(a)
+        assert a is None
 
         a = env.get_config_value('a', 42)
-        self.assertEqual(a, 42)
+        assert a == 42
 
 
 if __name__ == '__main__':
```
Comment (vishav1771) [inline tests/test_environment.py]: Remove 
```
if __name__ == '__main__':
 unittest.main()
```

File: tests/test_environment.py (hunk 2 above)
Comment (kidrahahjo) [reply tests/test_environment.py]: Thanks for the advice, will do it.

**Comment by vishav1771** [comment]
@csadorf @vyasr @bdice  Shouldn't we install `requirements-test.txt` for the testing? 
The error in the CI is due to click module which is present in `requirements-test.txt` but not in  `requirements.txt` and we are using it in the tests.

**Comment by kidrahahjo** [comment]
@vishav1771 
There still remains a problem
Error Message:
AttributeError: type object 'TestProject' has no attribute 'init_project'


Screenshot from 2020-02-01 01-37-27

**Comment by vishav1771** [comment]
@kidrahahjo  Change the name of  classes `TestProject` and `TestDynamicProject` present in `generate_test_project.py` to `ProjectTest` and `DynamicProjectTest`

**Comment by kidrahahjo** [comment]
@vishav1771 
<b>Errors in test_templates.py</b>
test_templates.py:95: in <module>
test_cls = type(test_name, TestBaseTemplate, dict(env=env))
TypeError: type.__new__() argument 2 must be tuple, not type

<b>Errors in test_project.py</b>
1. Attribute Error
self = <test_project.TestExecutionProject object at 0x7f51cd473b00>
def test_run(self):
with self.subTest(order='invalid-order'):
AttributeError: 'TestExecutionProject' object has no attribute 'subTest'
test_project.py:667: AttributeError

2. Using fixtures directly
Fixture "setUp" called directly. Fixtures are not meant to be called directly
but are created automatically when test functions request them as parameters.
See https://docs.pytest.org/en/latest/fixture.html for more information about fixtures, and
https://docs.pytest.org/en/latest/deprecations.html#calling-fixtures-directly about how to update your code.

3. Same Assertion Error inline 957

**Comment by vishav1771** [comment]
> 1. Attribute Error
>    self = <test_project.TestExecutionProject object at 0x7f51cd473b00>
>    def test_run(self):
>    with self.subTest(order='invalid-order'):
>    AttributeError: 'TestExecutionProject' object has no attribute 'subTest'
>    test_project.py:667: AttributeError

`subTest` is a method of `unittest` to remove this you have to add `pytest-subtest` to      
 requirements-dev.txt and requirements-test.txt and replace `self.subTest` with `pytest.subtests.test`

File: tests/test_project.py [hunk 4]
```diff
@@ -1007,40 +1007,42 @@ def op1(job):
...
+class TestProjectMainInterface(TestBaseProject):
+    project_class = ProjectTest
 
     def switch_to_cwd(self):
         os.chdir(self.cwd)
 
-    def setUp(self):
-        super(ProjectMainInterfaceTest, self).setUp()
+    @pytest.fixture(autouse=True)
+    def setUp(self,request):
```
Comment (vishav1771) [inline tests/test_project.py]:
```suggestion
    def setup_main_interface(self, request, setUp):
```
Add remove the next line

**Comment by kidrahahjo** [comment]
> > 1. Attribute Error
> >    self = <test_project.TestExecutionProject object at 0x7f51cd473b00>
> >    def test_run(self):
> >    with self.subTest(order='invalid-order'):
> >    AttributeError: 'TestExecutionProject' object has no attribute 'subTest'
> >    test_project.py:667: AttributeError
> 
> `subTest` is a method of `unittest` to remove this you have to add `pytest-subtest` to
> requirements-dev.txt and requirements-test.txt and replace `self.subTest` with `pytest.subtests.test`

Yes I tried implementing it after reading the pydocs but it didn't work for me

File: tests/test_templates.py [hunk 5]
```diff
@@ -84,17 +84,14 @@ def test_get_TestEnvironment(self):
...
+            assert '\n'.join(reference) == '\n'.join(generated)
 
 
 # TestCase factory
 for name, env in flow.environment.ComputeEnvironment.registry.items():
     if env.__module__.startswith('flow.environments'):
         name = '{}.{}'.format(env.__module__, env.__name__)[len('flow.'):]
         test_name = '{}TemplateTest'.format(env.__name__)
-        test_cls = type(test_name, (BaseTemplateTest, unittest.TestCase), dict(env=env))
+        test_cls = type(test_name, TestBaseTemplate, dict(env=env))
```
Comment (vishav1771) [inline tests/test_templates.py]:
```suggestion
        test_cls = type(test_name, (TestBaseTemplate, ), dict(env = env))
```

**Comment by kidrahahjo** [comment]
> > Yes I tried implementing it after reading the pydocs but it didn't work for me
> 
> To use `pytest.subtests` method you need to install `pytest-subtests` first. If you have done that already, can you share the error occurred using the above function

I guess I've overcome this problem by using subtests as a parameter for the function test_run

**Comment by kidrahahjo** [comment]
The **StringIO** error (line 958) still persists along with the **setUp** error

**Comment by vishav1771** [comment]
@kidrahahjo  Can you push your changes?

File: tests/test_project.py [hunk 6]
```diff
@@ -1036,7 +1037,7 @@ def switch_to_cwd(self):
         os.chdir(self.cwd)
 
     @pytest.fixture(autouse=True)
-    def setUp(self,request):
+    def setup_main_interface(self,request,setUp):
         super(TestProjectMainInterface, self).setUp()
```
Comment (vishav1771) [inline tests/test_project.py]: Remove this line 
```suggestion

````

Comment (kidrahahjo) [reply tests/test_project.py]: Yes I've tried this, it's not working too

Comment (kidrahahjo) [reply tests/test_project.py]: Okay now I've overcome this error too by no trying to add finalizer in my last line

**Comment by kidrahahjo** [comment]
<b>Issues Remaining</b>
1. test_templates.py
Screenshot from 2020-02-02 16-11-59

2. test_project.py
Screenshot from 2020-02-02 16-11-03

**Comment by kidrahahjo** [comment]
> @kidrahahjo Thank you very much for the contribution, I have a few comments/questions:
> 
> * Did you use [unittest2pytest](https://pypi.org/project/unittest2pytest/) or a similar tool to make these changes?
> * When I look at the [CI output](https://circleci.com/gh/glotzerlab/signac-flow/6657?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), it appears that there is a massive amount of data dumped to STDOUT. Do you have any idea what the issue could be?

@csadorf 
I've analyzed the data dumped and the possible solution to fix this issue is to fix all the code which will get deprecated in the coming versions as it's throwing warning error and returning the function where the error is coming from.

File: tests/.gitignore [hunk 7]
```diff
@@ -1,2 +0,0 @@
-compiled_scripts/
-template_reference_data/
```
Comment (csadorf) [inline tests/.gitignore]: Why was this file deleted?

File: tests/define_test_project.py [hunk 8]
```diff
@@ -3,32 +3,32 @@
 from flow import FlowProject
 
 
-class TestProject(FlowProject):
+class ProjectTest(FlowProject):
```
Comment (csadorf) [inline tests/define_test_project.py]: This class name is misleading, please rename to `_TestProject` and apply the same pattern wherever else this applies.

File: tests/test_project.py [hunk 9]
```diff
@@ -107,12 +107,13 @@ def is_present(cls):
         return True
 
 
-class BaseProjectTest(unittest.TestCase):
+class TestBaseProject():
```
Comment (csadorf) [inline tests/test_project.py]: This class name is misleading.

- [ ] Please rename `TestBaseProject` to `TestProjectBase`.

File: tests/test_project.py [hunk 10]
```diff
@@ -314,22 +315,22 @@ class A(FlowProject):
...
+            assert os.getcwd() == job.ws
 
         project = self.mock_project()
         with add_cwd_to_environment_pythonpath():
             with switch_to_directory(project.root_directory()):
                 starting_dir = os.getcwd()
                 with redirect_stderr(StringIO()):
                     A().run()
-                self.assertTrue(os.getcwd(), starting_dir)
+                assert os.getcwd(), starting_dir
```
Comment (csadorf) [inline tests/test_project.py]: Looks to me like this test had originally been incorrect. This should probably be:
```suggestion
                assert os.getcwd() == starting_dir
```

File: tests/test_project.py [hunk 11]
```diff
@@ -913,50 +915,49 @@ def test_submit_status(self):
...
+        assert len(list(MockScheduler.jobs())) == 0
         cluster_job_id = project._store_bundled(operations)
         stderr = StringIO()
         with redirect_stderr(stderr):
             project.submit_operations(_id=cluster_job_id, operations=operations)
-        self.assertEqual(len(list(MockScheduler.jobs())), 1)
-        self.assertIn('Some of the keys provided as part of the directives were not '
-                      'used by the template script, including: bad_directive\n',
-                      stderr.getvalue())
+        print(stderr.getvalue())
```
Comment (csadorf) [inline tests/test_project.py]: Please remove left-over print statement.

File: tests/test_templates.py [hunk 12]
```diff
@@ -16,12 +14,16 @@
 from test_project import redirect_stdout, redirect_stderr
 
 
-class BaseTemplateTest(object):
+class TestBaseTemplate(object):
```
Comment (csadorf) [inline tests/test_templates.py]: Please rename this class to `TestTemplateBase`.

File: tests/test_templates.py [hunk 13]
```diff
@@ -16,12 +14,16 @@
...
     project_class = signac.Project
-    env = None
+    envm = None
```
Comment (csadorf) [inline tests/test_templates.py]: Is there a good reason for this name change? If not, please revert it.

File: tests/test_templates.py [hunk 14]
```diff
@@ -84,17 +86,4 @@ def test_get_TestEnvironment(self):
                         with open(job.fn('script_{}.sh'.format(op))) as file:
                             reference.extend([msg] + file.read().splitlines())
 
-            self.assertEqual('\n'.join(reference), '\n'.join(generated))
-
-
-# TestCase factory
```
Comment (csadorf) [inline tests/test_templates.py]: Since you removed the test case factory, should this be individual sub tests now?

File: tests/test_templates.py [hunk 15]
```diff
@@ -16,12 +14,16 @@
 from test_project import redirect_stdout, redirect_stderr
 
 
-class BaseTemplateTest(object):
+class TestBaseTemplate(object):
     project_class = signac.Project
-    env = None
+    envm = None
+
+    for name, env in flow.environment.ComputeEnvironment.registry.items():
```
Comment (vishav1771) [inline tests/test_templates.py]: This loop should be inside your test function `test_get_TestEnvironment` and use `subtests.test` for every environment.

File: tests/test_templates.py [hunk 16]
```diff
@@ -33,8 +35,8 @@ def test_get_TestEnvironment(self):
         with signac.TemporaryProject(name=gen.PROJECT_NAME) as p:
             fp = gen.get_masked_flowproject(p)
             fp.import_from(origin=gen.ARCHIVE_DIR)
+            jobs = fp.find_jobs(dict(environments=self.env_name()))
```
Comment (vishav1771) [inline tests/test_templates.py]:
```suggestion
            jobs = fp.find_jobs(dict(environment=self.env_name()))
```
I think this is causing the error. Because of wrong argument it can't find any jobs.

File: tests/.gitignore (hunk 7 above)
Comment (kidrahahjo) [reply tests/.gitignore]: That was a severe mistake from my side, sorry for that.

File: tests/test_templates.py (hunk 13 above)
Comment (kidrahahjo) [reply tests/test_templates.py]: Actually, no, I will correct it right away.

**Comment by csadorf** [comment]
@kidrahahjo Can you either rebase or merge origin/master?

File: tests/test_templates.py [hunk 17]
```diff
@@ -33,8 +31,8 @@ def test_get_TestEnvironment(self):
         with signac.TemporaryProject(name=gen.PROJECT_NAME) as p:
             fp = gen.get_masked_flowproject(p)
             fp.import_from(origin=gen.ARCHIVE_DIR)
-
             jobs = fp.find_jobs(dict(environment=self.env_name()))
+
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply tests/test_templates.py]: I'm sorry I didn't quite understand the issue over here?

**Comment by kidrahahjo** [comment]
> @kidrahahjo Can you either rebase or merge master?

I'll do it.

File: tests/test_templates.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (csadorf) [reply tests/test_templates.py]: It looks like you moved code on line 34 resulting in two changes (a line removal and an empty line added). I'd prefer to keep the PR free of unrelated changes like this, but it's not a big deal.

File: tests/test_templates.py [hunk 18]
```diff
@@ -18,24 +19,31 @@ class TestTemplateBase(object):
...
         return '.'.join(name.split('.')[1:])
 
-    def get_TestEnvironment(self):
+    def find_envs():
+        for name, env in flow.environment.ComputeEnvironment.registry.items():
+            if env.__module__.startswith('flow.environments'):
+                yield env
+
+    @pytest.mark.parametrize('env', find_envs())
+    def test_env(self, env):
```
Comment (csadorf) [inline tests/test_templates.py]: This should be a module-level function now, there is no need to have a test class anymore.

Comment (kidrahahjo) [reply tests/test_templates.py]: If I don't use self, this error comes up.
In test_env: function uses no argument 'env'

Comment (kidrahahjo) [reply tests/test_templates.py]: What I can do is, remove the class and just let test_templates.py to be a set of functions.

Comment (csadorf) [reply tests/test_templates.py]: Yes, exactly, that's what I meant by "not have a test class".

**Comment by kidrahahjo** [comment]
Yes I just read it carefully, having a bad day I guess. Sorry about that.

File: tests/test_templates.py (hunk 18 above)
Comment (kidrahahjo) [reply tests/test_templates.py]: Yes I just read it carefully, having a bad day I guess. Sorry about that.

**Comment by kidrahahjo** [comment]
> You did not properly resolve all merge conflicts. Please follow instructions here: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line

There are no merge conflicts, and it says your branch is up to date with origin master

**Comment by csadorf** [comment]
> > You did not properly resolve all merge conflicts. Please follow instructions here: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line
> 
> There are no merge conflicts, and it says your branch is up to date with origin master

Please see for example the diff here which contains incorrectly resolved conflicts: https://github.com/glotzerlab/signac-flow/pull/232/files#diff-e8800259c4985f4a73b0936128034c53

**Comment by kidrahahjo** [comment]
> In the latest revision, all changes to the changelog have now completely disappeared which means that your third attempt at synchronizing with master has gone somewhat wrong as well.
> 
> I must admit that I am starting to get a bit frustrated. Resolving conflicts with the base branch is not a trivial task so I fully understand if that is a difficult process. There is zero shame in having little experience in handling merge conflicts and in asking for help with that. However, I'd urge you to be a bit more diligent and dismissing my comments is not very respectful of me and my time.
> 
> In an effort to support you, I have checked out your branch and performed the merge myself into the last good commit ([c4f7990](https://github.com/glotzerlab/signac-flow/commit/c4f79900ee07f1f86a618ba0f216ee06fa729e80)). I can push it to this branch if you like, however, you will need to provide me with the permissions to do so: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

With all due respect sir, I already gave the permissions to do so and I'm sorry if I dismissed any comment of yours. 
I may not be aware of how and when to use resolve conversation button, whenever you suggested a change of code and I changed that particular piece, I resolved the conversation thinking that I've made the changes you told me to.

**Comment by kidrahahjo** [comment]
> LGTM!  
> 
> I would still ask the other reviewers to carefully review the change set, but I am happy with it now.
> 
> @kidrahahjo Thank you very much for taking on this huge task and accomplishing it so quickly. The switch to pytest will significantly simplify the implementation and maintenance of tests in the future. We really appreciate your efforts.

Thank you sir, means a lot. Will definitely keeping contributing in this wonderful community.

File: tests/test_project.py [hunk 19]
```diff
@@ -1007,40 +1006,41 @@ def op1(job):
...
+    project_class = _TestProject
 
     def switch_to_cwd(self):
         os.chdir(self.cwd)
 
-    def setUp(self):
-        super(ProjectMainInterfaceTest, self).setUp()
+    @pytest.fixture(autouse=True)
+    def setup_main_interface(self, request):
+        # super(TestProjectMainInterface, self).setUp()
```
Comment (bdice) [inline tests/test_project.py]: Remove commented line.
```suggestion
```

File: tests/test_project.py [hunk 20]
```diff
@@ -1052,71 +1052,71 @@ def call_subcmd(self, subcmd):
                 with switch_to_directory(self.project.root_directory()):
                     return subprocess.check_output(_cmd.split(), stderr=subprocess.DEVNULL)
         except subprocess.CalledProcessError as error:
-            print(error, file=sys.stderr)
-            print(error.output, file=sys.stderr)
+            print(error, sys.stderr)
```
Comment (bdice) [inline tests/test_project.py]: I would prefer to revert these two lines to include `file=`.
```suggestion
            print(error, file=sys.stderr)
```

File: tests/test_project.py [hunk 21]
```diff
@@ -1052,71 +1052,71 @@ def call_subcmd(self, subcmd):
...
+            print(error.output, sys.stderr)
```
Comment (bdice) [inline tests/test_project.py]:
```suggestion
            print(error.output, file=sys.stderr)
```

File: requirements-dev.txt [hunk 22]
```diff
@@ -1,3 +1,5 @@
-click>=7.0
+Click>=7.0
```
[the remark this answers is not in the document]
Comment (csadorf) [reply requirements-dev.txt]: @bdice Not sure why the change, but just so you know: PyPI (pip) is case insensitive: https://www.python.org/dev/peps/pep-0426/#name

**Comment by csadorf** [comment]
@kidrahahjo Just to make sure you are aware: I needed to force-push my changes to this branch. So please reset your local branch with `git reset origin/master --hard` before you make any new commits.

File: requirements-dev.txt (hunk 22 above)
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply requirements-dev.txt]: When I typed pip install click in the cimmand line it showed me an error. And it resolved when Click was typed instead of click. I guess there was a typo error if pip is case insensitive. I'll change it if you want.

File: tests/test_project.py [hunk 23]
```diff
@@ -913,50 +914,48 @@ def test_submit_status(self):
...
         stderr = StringIO()
         with redirect_stderr(stderr):
             project.submit_operations(_id=cluster_job_id, operations=operations)
-        self.assertEqual(len(list(MockScheduler.jobs())), 1)
-        self.assertIn('Some of the keys provided as part of the directives were not '
-                      'used by the template script, including: bad_directive\n',
-                      stderr.getvalue())
+        assert len(list(MockScheduler.jobs())) == 1
+        assert 'Some of the keys provided as part of the directives were not used by the template '
+        'script, including: bad_directive' in stderr.getvalue()
```
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply tests/test_project.py]: It was just the logic that if 'a' is in 'a\nbc' then what's the point of using 'a\n'

File: requirements-dev.txt [hunk 24]
```diff
@@ -1,3 +1,4 @@
 click>=7.0
 ruamel.yaml>=0.15.89
 flake8>=3.7.7
+
```
Comment (csadorf) [inline requirements-dev.txt]: Please remove the extraneous line.

As a tip: Please use `git add -up` and not just `git add` in combination with `git diff --staged` whenever you prepare your commits. In that way you are prompted to confirm and review every single chunk of change you introduce making it so that errors like this are much less likely to occur.

File: requirements.txt [hunk 25]
```diff
@@ -2,3 +2,5 @@ signac>=1.0.0
 jinja2>=2.8
 cloudpickle
 deprecation>=2
+pytest>=4.4
+pytest-subtests
```
Comment (csadorf) [inline requirements.txt]: These packages are *not* needed to run the package, just for testing.

File: requirements-test.txt [hunk 26]
```diff
@@ -1,5 +1,5 @@
-Click==7.0
+click==7.0
 ruamel.yaml==0.15.89
 flake8==3.7.7
-pytest>=3.5.0
-pytest-subtests==0.3.0
+pytest>=4.4
```
Comment (csadorf) [inline requirements-test.txt]: Just a question: Is there a specific reason that you choose version 4.4?

Comment (kidrahahjo) [reply requirements-test.txt]: Actually while using pytest-subtest the minimum requirement is [Pytest Version 4.4](https://github.com/pytest-dev/pytest-subtests)

File: requirements-dev.txt (hunk 24 above)
Comment (kidrahahjo) [reply requirements-dev.txt]: Sure, thanks

File: requirements-test.txt (hunk 26 above)
Comment (csadorf) [reply requirements-test.txt]: Got it, thank you.

**Comment by bdice** [comment]
@kidrahahjo Can you go back through and click "Resolve" on the things you fixed from my previous review? I am not sure what's left to be done, if anything.

File: tests/test_project.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (kidrahahjo) [reply tests/test_project.py]: I made the changes but I have one doubt that why was it required?

**Comment by kidrahahjo** [comment]
> @kidrahahjo Can you go back through and click "Resolve" on the things you fixed from my previous review? I am not sure what's left to be done, if anything.

done

File: tests/test_project.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (bdice) [reply tests/test_project.py]: @kidrahahjo We should not change the test itself while we change the testing framework. The previous test had a newline, so the new test should match exactly. There were only very minor exceptions to this in the rest of the PR, addressing tests that were incorrect in the previous code or written in a confusing way (like the one that had extra parentheses).

As for why this newline is needed in the first place: we could have other unused directives. This is a case where the presence of the newline would affect the test's behavior:
```
Some of the keys provided as part of the directives were not used by the template used by the template script, including: bad_directive, some_other_directive
```
Note that the string without a newline is a substring of this message, but the string with a newline is not a substring.

Please make that change and then I can approve this PR.

[the remark this answers is not in the document]
Comment (kidrahahjo) [reply tests/test_project.py]: @bdice Got it, thanks. I have made the change already, just because I had the doubt I didn't resolve this conversation

**Comment by csadorf** [comment]
@vishav1771 Do you want to have a last look at this before we merge?

**Comment by bdice** [comment]
I feel confident that this PR can be merged without needing a third approval from @vishav1771. In the interest of not blocking #244 or introducing conflicts with other PRs, I will merge this. Thanks again for your help, @kidrahahjo @vishav1771 @csadorf! 🎉📈
